[NM-82] download results state refactor

M-Kusumgar opened the pull request:
## Description

This one was painful. Ill remove AGYW in the next PR to see how easy/hard it is to modify this new construction, almost 2000 lines gone, damn.

There was a handy enum in the `types` file called `DownloadType` so i thought way not roll with that, everything is constructed from `downloadConfig.ts`, using `Object.values` on a string enum (does not work well on a usual enum, can try it if you want to know a stupid typescript decision) gives all the values and we can loop over it to do everything

The tests were especially painful, i suspect there'll be small simplifications when we remove AGYW

The other thing is I had to add a payload handler to the api service because i needed finer control over the payload the api need to commit to the state (needed to add `type` so that it knows which `DownloadType` to commit to)

ideally, to add another button you just need to extend the 3 things in `downloadConfig.ts`

Edit: the e2e test failure is a real one! i am removing AGYW in #1037 which im merging into here so will just fix the e2e test there

r-ash commented on src/app/static/src/app/components/downloadResults/DownloadResults.vue:
```diff
@@ -2,47 +2,17 @@
-                    <error-alert v-if="spectrum.downloadError" :error="spectrum.downloadError"></error-alert>
+                <template v-for="type in Object.values(DownloadType)" :key="type">
+                    <div :id="`${type}-download`" v-if="switches[type]">
+                        <download :translate-key="`${type}Download`"
```
I wonder if we should put this into the `downloadConfig.ts` like the `switches` a map of type to the key we want to use. I do like all the download stuff being in the single config file.

M-Kusumgar replied:
hmmm do you mean because theres duplication between this and the `locales.ts` type? because otherwise everything is auto generated, also the types in the locales file ensures we have the matching translation key

r-ash replied:
No, only that when you're adding a new file type you have to look here to see what the key is. It is `{type}Download" instead of just being able to see the full string in the `downloadConfig.ts` ? That make sense?

r-ash commented on src/app/static/src/app/store/adrUpload/actions.ts:
```diff
@@ -111,23 +112,23 @@ export const actions: ActionTree<ADRUploadState, RootState> & ADRUploadActions =
             }
             if (resourceType === rootState.adr.schemas?.outputSummary) {
-                downloadId = rootState.downloadResults.summary.downloadId
+                downloadId = rootState.downloadResults[DownloadType.SUMMARY].downloadId
```
What do you also think about handling this through the `downloadConfig`? Have some list of files which need to go to the ADR, and here look through the `DownloadType` and include those which are in the list?

r-ash replied:
Could use it above in `UploadModal`

M-Kusumgar replied:
the reason i didnt do this here is because this requires me to refactor the ADR upload stuff because we need some way of matching up if "outputSummary" exists, then include `DownloadType.SUMMARY`, i say we do a smaller refactor in a separate PR and sort out the adrUpload stuff and there we can put these mappings, infact some of the adrUpload stuff can probably just come from the downloadConfig itself

in the `UploadModal` we have this same mapping issue and also have `outputFileError` which has a weird control flow which does something different if both spectrum and summary have metadata errors vs just one of them, i would like to properly investigate that and see if theres a pattern that we can extract

i can create a ticket for these two if you agree!

r-ash replied:
Yep sounds great doing that separately, yes please create the tickets.

r-ash commented on src/app/static/src/app/store/root/mutations.ts:
```diff
@@ -141,6 +136,12 @@ export const mutations: MutationTree<RootState> = {
 
 };
 
+const stopPollingAllDownloadTypes = (state: RootState) => {
```
:heart:

r-ash commented on src/app/static/src/app/apiService.ts:
```diff
@@ -70,9 +70,10 @@ export class APIService<S extends string, E extends string> implements API<S, E>
 
-    withError = (type: E, root = false) => {
+    withError = (type: E, root = false, payloadFunc?: (e: Error) => any) => {
         this._onError = (failure: Response) => {
-            this._commit({type: type, payload: APIService.getFirstErrorFromFailure(failure)}, {root});
+            const error = APIService.getFirstErrorFromFailure(failure);
+            this._commit({ type, payload: payloadFunc ? payloadFunc(error) : error }, { root });
```
So this is what I had previously added `withErrorCallback` for. But I think what you've done is better. Could you remove the `withErrorCallback` helper and tidy up uses of that to use this instead?

But actually, I wonder if we should have 2 functions `withError` and `withErrorHandler`? Or some better name? Whatever you think!

M-Kusumgar replied:
ohhhh damn sorry i didnt even see that, i was looking for that type of functionality, yh im happy to remove it, i think its nice to declare the mutation you want and all that, i personally think its nicer if we just have one function with an optional arg, it feels like exactly the same responsibility, just with pre commit step or not

r-ash replied:
Nope that's fine, happy with 1 function if you want.

r-ash reviewed (commented):
This looks really great, so good to see so much duplication removed.

Couple of comments on a few bits would be nice to get your thoughts on. But I think this looks good to go

M-Kusumgar commented on src/app/static/src/app/store/downloadResults/actions.ts:
```diff
@@ -3,273 +3,83 @@ import {RootState} from "../../root";
 import {DownloadResultsMutation} from "./mutations";
-import {ModelStatusResponse} from "../../generated";
-import {DOWNLOAD_TYPE} from "../../types";
+import {Error, ModelStatusResponse} from "../../generated";
```
yep removed, neither was switches before it! you can tell i did a couple of drafts before settling on the final version haha

M-Kusumgar commented on src/app/static/src/app/store/downloadResults/downloadConfig.ts:
```diff
@@ -0,0 +1,31 @@
+// we do posts to /download/submit/<url>/<calibrate-id> with the body function called
+type PostObject = { url: string, body: (store: ActionContext<DownloadResultsState, RootState>) => object };
+export const downloadTypePrepareUrls: Record<DownloadType, PostObject> = {
+    [DownloadType.SPECTRUM]: { url: "spectrum", body: (store) => store.rootGetters.projectState },
+    [DownloadType.COARSE]: { url: "coarse-output", body: () => ({}) },
+    [DownloadType.SUMMARY]: { url: "summary", body: () => ({}) },
+    [DownloadType.COMPARISON]: { url: "comparison", body: () => ({}) },
+    [DownloadType.AGYW]: { url: "agyw", body: () => ({}) }
```
agreed, in fact lets do a bit more generally, lets just make both the body and url functions that take in the store, that would be good!

M-Kusumgar replied:
done

M-Kusumgar commented on src/app/static/src/app/store/downloadResults/mutations.ts:
```diff
@@ -1,251 +1,96 @@
-    AgywError = "AgywError",
-    AgywStatusUpdated = "AgywStatusUpdated",
-    AgywDownloadError = "AgywDownloadError",
+    Preparing = "Preparing",
```
yh these were definitely satisfying files!

M-Kusumgar commented:
have addressed comments in this commit: https://github.com/hivtools/hint/pull/1037/commits/481b209187f025a78c929ac990875417017ee64d

ADR upload refactor ticket is created!: https://www.notion.so/unaids-naomi/Upload-ADR-state-refactor-f1c71f2f341c49aa98bb9f3bad88e99c?pvs=4

r-ash reviewed (approved):
Thanks for that, I just merged in #1034 and then removed `downloadOutput` action. We don't need that any more as we're now downloading via an href instead of an action
